linkerd/system/src/linux.rs

```diff
@@ -39,9 +42,28 @@ pub fn open_fds(pid: pid_t) -> io::Result<u64> {
 }
 
 pub fn max_fds() -> io::Result<Option<u64>> {
-    let limit = pid::limits_self()?.max_open_files;
-    let max_fds = limit.soft.or(limit.hard).map(|max| max as u64);
-    Ok(max_fds)
+    match Process::myself() {
+        Ok(p) => match p.limits() {
+            Ok(limits) => {
+                let limit = limits.max_open_files;
+                match limit.soft_limit {
+                    process::LimitValue::Unlimited => match limit.hard_limit {
+                        process::LimitValue::Unlimited => Ok(None),
+                        process::LimitValue::Value(v) => Ok(Some(v)),
+                    },
+                    process::LimitValue::Value(v) => Ok(Some(v)),
+                }
+            }
+            Err(e) => {
+                error!("Failed to get process limits: {}", e);
+                Ok(None)
+            }
+        },
+        Err(e) => {
+            error!("Failed to get process limits: {}", e);
+            Ok(None)
```

Comment:
It seems that the caller of this function can handle three distinct types of values:

1. Ok(Some)
2. Ok(None) (do nothing)
3. Err, where we log the error we've encountered.

Should we perhaps return the error instead of returning Ok(None)? Feels a bit weird to log here, at a different level than the caller, when the caller already handles this path?